src/computational_graph/graph.jl

```diff
@@ -85,10 +49,11 @@ mutable struct Graph{F,W} # Graph
 
     external::Vector{Int} # index of external vertices
     vertices::Vector{OperatorProduct} # vertices of the diagram
+    topology::Vector{Vector{Int}}
 
     subgraph::Vector{Graph{F,W}}
 
-    operator::Operator
+    operator::DataType
```

Comment:
I think this type definition is confusing, because `operator` is a concrete subtype of abstract type Operator, not a generic concrete DataType. Why not try [union splitting](https://julialang.org/blog/2018/08/union-splitting/)? (see also: [here](https://github.com/JuliaLang/julia/pull/44131))

In particular, we would rename abstract type `Operator` to `AbstractOperator`, and define a new concrete union type like `const Operator = Union{Sum,Prod,...}` which is both extensible and self-documenting, and then  we have a type-stable `operator::Operator` for the field in struct `Graph`.

The performance for large unions may not be generically better than the current solution in all cases, but it seems this approach has improved a lot as of Julia 1.7 and the code would be clearer, so we should try it and profile.